### PR TITLE
fix: speed up parsing long lists

### LINKED ISSUE #2297: Performance of parse() is 15 times worse in releases after 2.1.3
**Marked version: >2.1.3**

**Describe the bug**
The speed of parse() call with large markdown source was about 15 times better in v2.1.3 than releases after it. Maximum size of markdown source that can be processed in the browser is respectively way smaller.

**To Reproduce**

const startTime = Date.now();
marked.parse(mdData800kB);
console.log('time in milliseconds', Date.now() - startTime);

Output:
time in milliseconds 30000

**Expected behavior**

time in milliseconds 2000

### PR DESCRIPTION
**Pull request by calculuschild**
**Marked version:**

v3.0.0 +

## Description

- Fixes #2297. The rework of Lists in v3.0.0 accidentally introduced a O(n^2) slowdown issue for long documents.

The List tokenizer was using a RegEx to capture the potential next list Item, and then split that captured text to work line-by-line to determine if it had proper indentation, etc. and whether each line should be part of the current list Item.

Problem is, the captured text was literally *the entire document*, so for every potential list item, we were capturing the entire document and then splitting the document into lines. For longer documents, this meant spending the majority of time just splitting the document into lines over and over.

Here's the offending Regex, and notice that `(?:\\n[^\\n]*)*` matches *everything to the end of the document*:

https://github.com/markedjs/marked/blob/d098d554894ac4eeb4f7226dd80a5daebd59911b/src/Tokenizer.js#L193

This PR changes that so we only capture the first line (with it's bullet point), and once we verify that it is a candidate for starting a new list Item, we just traverse the SRC one line at a time. No more mass line-splitting when we really only need to look one line at a time anyway.

Offending line-splitting line where 90% of processing time was spent:

https://github.com/markedjs/marked/blob/d098d554894ac4eeb4f7226dd80a5daebd59911b/src/Tokenizer.js#L205

Needs a bit of cleanup, but it's passing tests.

## Contributor

- [ ] Test(s) exist to ensure functionality and minimize regression (if no tests added, list tests covering this PR); or,
- [x] no tests required for this PR.
- [ ] If submitting new feature, it has been documented in the appropriate places.

## Committer

In most cases, this should be a different person than the contributor.

- [ ] CI is green (no forced merge required).
- [ ] Squash and Merge PR following [conventional commit guidelines](https://www.conventionalcommits.org/).

### CODE REVIEW

File: test/specs/bug/adjacent_lists.md
Comment (UziTech) [inline test/specs/bug/adjacent_lists.md:1]: Why was this test removed? Is it failing with this PR?

Comment (UziTech) [reply test/specs/bug/adjacent_lists.md:1]: nvm I see this was an accidental addition.

Comment (calculuschild) [reply test/specs/bug/adjacent_lists.md:1]: That was a test I added playing around trying to debug my program and accidentally committed it to the PR.

File: test/specs/redos/quadratic_lists.cjs
Comment (UziTech) [inline test/specs/redos/quadratic_lists.cjs:4]: I can confirm this speeds up parsing long lists.

This test takes about 2 seconds on my machine on master and about 100 ms with this PR.

**Review by UziTech — APPROVED** [review]

File: test/specs/redos/quadratic_lists.cjs
Comment (calculuschild) [reply test/specs/redos/quadratic_lists.cjs:4]: Hooray!

**Comment by calculuschild** [comment]
I want to go over this code once more before it gets merged. I have a nagging sense that there's still some debris left behind I should clean up but exactly where is eluding me at the moment.

File: lib/marked.cjs
Comment (styfle) [inline lib/marked.cjs:564]: Should this comment be removed?

File: lib/marked.cjs
Comment (styfle) [inline lib/marked.cjs:668]: Also here, should this comment be removed?

File: lib/marked.cjs
Comment (calculuschild) [reply lib/marked.cjs:564]: This was leftover in the compiled file looks like. Recompiled so this should now be up to date.

**Comment by calculuschild** [comment]
@Uzitech Cleaned up the logic how I wanted. Passes the specs but is failing the SNYK security test here and I'm not sure why.

Otherwise, this is now ready to merged.

File: src/Tokenizer.js
Comment (calculuschild) [inline src/Tokenizer.js:305]: Do list items need to be marked as loose or just the list itself? This is a spot we could speed up slightly if we keep it commented out; it doesn't fail any tests.

Comment (UziTech) [reply src/Tokenizer.js:305]: It looks like the parser only uses the `loose` property on the list not the list items. I think it would be fine to not have this on the list items but it might be considered a breaking change.

Comment (calculuschild) [reply src/Tokenizer.js:305]: I will leave that up to you then. If needed we can lump it in with other breaking changes at a later date?

Comment (UziTech) [reply src/Tokenizer.js:305]: Actually it looks like that property is not necessarily correct in the list items anyway. Only the list items with a `space` token have the loose property set to true but it should be all items if any of them have a `space` token.

Comment (UziTech) [reply src/Tokenizer.js:305]: Ya I would say leave it right now and we can remove it later.

Comment (calculuschild) [reply src/Tokenizer.js]: Ok, it's uncommented now.

> Actually it looks like that property is not necessarily correct in the list items anyway

Ah yeah, you're right. I guess that mistake didn't get caught because it's never used anywhere.

**Review by UziTech — APPROVED** [review]

File: lib/marked.esm.js
Comment (styfle) [inline lib/marked.esm.js:588]: Remove this comment too?

File: lib/marked.umd.js
Comment (styfle) [inline lib/marked.umd.js:672]: Another comment that seems out of place here

File: lib/marked.cjs
Comment (styfle) [inline lib/marked.cjs:670]: This one too

File: lib/marked.esm.js
Comment (calculuschild) [reply lib/marked.esm.js:588]: These are just the compiled files which are out of date. They are not present in the actual source code so I don't think they matter here (they get regenerated when we merge, right?), but I will recompile now to see what happens.

Comment (calculuschild) [reply lib/marked.esm.js]: There. Lib files were updated.

**Review by styfle — APPROVED** [review]

**Pull request merged**
